### PR TITLE
Make `axum-debug` handle more cases

### PR DESCRIPTION
**Pull request by davidpdrsn**
This is basically a rewrite of `axum-debug` to make it handle more cases. Sorry @programatik29 😞

Cases now supported:

```rust
// mut args
#[debug_handler]
async fn handler(mut foo: String) {
}

// returning `impl Future`
#[debug_handler]
fn handler() -> impl Future<Output = ()> {
    async {}
}

// returning named future
#[debug_handler]
fn handler() -> std::future::Ready<()> {
    std::future::ready(())
}

struct A;

impl A {
    // returning `Self`
    #[debug_handler]
    async fn handler() -> Self {
        A
    }

    // extracting `self`
    #[debug_handler]
    async fn handler(self) {
    }
}
```

The only case I was able to think off that I couldn't figure out was:

```rust
impl A {
    #[debug_handler]
    async fn handler() {
    }
}
```

This doesn't work because the macro can't know that `handler` must be called as `A::handler` and there is no way to know since it doesn't take `self` or return `Self`. The only way to know would be adding an attribute on the `impl` as well.

All previously supported cases are also supported but with slightly different error messages in some cases.

Fixes https://github.com/tokio-rs/axum/issues/514
Fixes https://github.com/tokio-rs/axum/issues/515

### CODE REVIEW

**Comment by davidpdrsn** [comment]
Actually I just noticed that generic handlers like `async fn foo<T: FromRequest>(t: T) {}` don't work. Gonna give that a shot.

**Comment by davidpdrsn** [comment]
Hm not sure generics can be supported easily 🤔 Its quite a can of worms. I changed it to just return an error if you're using the macro on a generic function.

File: axum-debug/src/lib.rs [hunk 1]
```diff
@@ -124,298 +124,247 @@
...
-                            receiver,
-                            "`#[debug_handler]` is not supported on methods",
-                        ));
-                    }
-                };
+    fn unique_name() -> Ident {
+        use std::sync::atomic::{AtomicU64, Ordering};
+        static COUNT: AtomicU64 = AtomicU64::new(0);
+        let count = COUNT.fetch_add(1, Ordering::SeqCst);
+        quote::format_ident!("__axum_debug_check_{}", count)
```
[the remark this answers is not in the document]
Comment (jplatte) [reply axum-debug/src/lib.rs:324]: This also makes the macro non-deterministic, which negatively impacts later stages of incr-comp... Not sure whether the idea is to leave these in after fixing the issue, but wouldn't it generally be better to use names like `format!("__axum_debug_{}_check_response", input_name)`?

**Comment by jplatte** [comment]
> This doesn't work because the macro can't know that `handler` must be called as `A::handler` and there is no way to know since it doesn't take `self` or return `Self`.

Couldn't you only generate the checking code alongside the input if there is a `self` or `Self` in the function signature (for `self` checking the first argument, for `Self` using the `visit` API) and generate an inner function as before if that is not the case?

File: axum-debug/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum-debug/src/lib.rs:324]: Uh very good point 🤔 Will fix.

**Comment by davidpdrsn** [comment]
> Couldn't you only generate the type assertion functions alongside the input if there is a self or Self in the function signature (for self checking the first argument, for Self using the visit API) and generate an inner function as before if that is not the case?

Thats a good idea! I'll work on that.

**Comment by davidpdrsn** [comment]
@jplatte Thanks for the good feedback! I think `debug_handler` should work everywhere now 🤞

File: axum-debug/src/lib.rs [hunk 2]
```diff
@@ -268,19 +268,33 @@ mod debug_handler {
...
+        } else {
+            quote_spanned! {span=>
+                #[allow(warnings)]
+                async fn #name() {
+                    #make
+
+                    let value = #make_value_name().await;
+                    fn check<T>(_: T)
+                        where T: ::axum::response::IntoResponse
+                        {}
```
Comment (jplatte) [inline axum-debug/src/lib.rs:295]:
```suggestion
                    {}
```

**Pull request merged**